internal/storage/deleter.go

```diff
@@ -0,0 +1,165 @@
+package storage
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/pkg/errors"
+)
+
+// Deleter removes a pod and its components from storage
+type Deleter struct {
+	db *sql.DB
+}
+
+// NewDeleter creates a new delete
+func NewDeleter(db *sql.DB) *Deleter {
+	return &Deleter{db}
+}
+
+// Delete a pod from storage
+func (d *Deleter) Delete(ctx context.Context, pod Pod) error {
+	txn, err := d.db.BeginTx(ctx, nil)
+	if err != nil {
+		return errors.Wrap(err, "failed to begin transaction")
+	}
+	defer txn.Rollback()
+
+	row := txn.QueryRowContext(ctx, `
+SELECT
+	id
+FROM
+	pods
+WHERE
+	name=$1
+	`, pod.Name)
+
+	var podID int64
+	err = row.Scan(&podID)
+	if err != nil {
+		return errors.Wrap(err, "failed to get pod")
+	}
+
+	deleteQuerys := []string{
+		"delete from pod_processors where pod=$1;",
+		"delete from pod_sources where pod=$1;",
+		"delete from pod_views where pod=$1;",
+		"delete from pod_sinks where pod=$1;",
+		"delete from pod_view_sinks where pod=$1;",
+		"delete from pod_view_sources where pod=$1;",
+		"delete from pods where id=$1;",
+	}
+
+	for _, query := range deleteQuerys {
+		_, err := txn.ExecContext(ctx, query, podID)
+		if err != nil {
+			return errors.Wrap(err, "failed to delete pod")
+		}
+	}
+
+	cleanupQuerys := []string{
+		`
+delete from
+	processor_inputs
+where not exists
+	(select 1 from pod_processors where pod_processors.processor=processor_inputs.processor);
+		`,
+		`
+delete from
+	processor_joins
+where not exists
+	(select 1 from pod_processors where pod_processors.processor=processor_joins.processor);
+		`,
+		`
+delete from
+	processor_lookups
+where not exists
+	(select 1 from pod_processors where pod_processors.processor=processor_lookups.processor);
+		`,
+		`
+delete from
+	processor_outputs
+where not exists
+	(select 1 from pod_processors where pod_processors.processor=processor_outputs.processor);
+		`,
+		`
+delete from
+	processors
+where not exists
+	(select 1 from pod_processors where pod_processors.processor=processors.id);
+		`,
+		`
+delete from
+	views
+where not exists
+	(select 1 from pod_views where pod_views.view=views.id);
+		`,
+		`
+delete from
+	sources
+where not exists
+	(select 1 from pod_sources where pod_sources.source=sources.id);
+		`,
+		`
+delete from
+	view_sinks
+where not exists
+	(select 1 from pod_view_sinks where pod_view_sinks.view_sink=view_sinks.id);
+		`,
+		`
+delete from
+	view_sources
+where not exists
+	(select 1 from pod_view_sources where pod_view_sources.view_source=view_sources.id);
+		`,
+		`
+delete from
+	topics
+where not exists
+	(
+		select 1 from sources where sources.topic=topics.id union
```

Comment:
TIL that postgres handles this The Right Way™️. The semantics of `UNION` change slightly within an `EXISTS` clause; the `SELECT` statement is ignored and the result of matching records is unioned and evaluated against the `WHERE` clause.

Admittedly, this nerd-sniped me and I wrote this SQL query to prove to myself that this works as intended:

```sql

WITH topic AS (
    INSERT INTO topics (name, message)
    VALUES ('t1', 'm1'), ('t2', 'm2'), ('t3', 'm3'), ('t4', 'm4')
    ON CONFLICT(name) DO NOTHING
    RETURNING name, id
),
service AS (
    INSERT INTO services (name, description)
    VALUES ('s1', 'blah')
    ON CONFLICT(name) DO NOTHING
    RETURNING name, id
),
component AS (
    INSERT INTO components (name, service, description)
    SELECT 'c1', s.id, 'blah'
    FROM service s
    WHERE s.name = 's1'
    ON CONFLICT(name, service) DO NOTHING
    RETURNING name, id
),
source AS (
    INSERT INTO sources (component, topic)
    SELECT c.id, t.id
    FROM component c, topic t
    WHERE c.name = 'c1'
    AND t.name = 't1'
    ON CONFLICT(component, topic) DO NOTHING
    RETURNING id
),
view AS (
    INSERT INTO views (component, topic)
    SELECT c.id, t.id
    FROM component c, topic t
    WHERE c.name = 'c1'
    AND t.name = 't2'
    ON CONFLICT(component, topic) DO NOTHING
    RETURNING id
)
SELECT *
FROM source, view;

SELECT * FROM topics;

delete from
	topics
where not exists
	(
		select 1 from sources where sources.topic=topics.id union
		select 1 from views where views.topic=topics.id union
		select 1 from view_sources where view_sources.topic=topics.id union
		select 1 from view_sinks where view_sinks.topic=topics.id union
		select 1 from processors where processors.persistence=topics.id union
		select 1 from processor_inputs where processor_inputs.topic=topics.id union
		select 1 from processor_joins where processor_joins.topic=topics.id union
		select 1 from processor_lookups where processor_lookups.topic=topics.id union
		select 1 from processor_outputs where processor_outputs.topic=topics.id union
		select 1 from sinks where sinks.topic=topics.id
	)
RETURNING *;

SELECT * FROM topics;
```